packages/service/ni_measurement_plugin_sdk_service/measurement/client_support.py

```diff
@@ -1,15 +1,96 @@
 """Support functions for the Measurement Plug-In Client."""
 
-from ni_measurement_plugin_sdk_service._internal.parameter.decoder import deserialize_parameters
-from ni_measurement_plugin_sdk_service._internal.parameter.encoder import serialize_parameters
-from ni_measurement_plugin_sdk_service._internal.parameter.metadata import ParameterMetadata
+from pathlib import Path
+from typing import Any, Dict, Iterable, List
+
+from ni_measurement_plugin_sdk_service._internal.parameter.decoder import (
+    deserialize_parameters,
+)
+from ni_measurement_plugin_sdk_service._internal.parameter.encoder import (
+    serialize_parameters,
+)
+from ni_measurement_plugin_sdk_service._internal.parameter.metadata import (
+    ParameterMetadata,
+)
 from ni_measurement_plugin_sdk_service._internal.parameter.serialization_descriptors import (
     create_file_descriptor,
 )
 
 __all__ = [
     "create_file_descriptor",
+    "convert_paths_to_strings",
+    "convert_strings_to_paths",
     "deserialize_parameters",
     "ParameterMetadata",
     "serialize_parameters",
 ]
+
+
+def convert_paths_to_strings(
+    parameter_metadata_dict: Dict[int, ParameterMetadata], parameter_values: Iterable[Any]
+) -> List[Any]:
+    """Convert path parameters from Path objects to strings.
+
+    Args:
+        parameter_metadata_dict: Parameter metadata by ID.
+
+        parameter_values: The parameter values.
+
+    Returns:
+        The parameter values with Path objects converted to strings, where appropriate.
+    """
+    result: List[Any] = []
+
+    for index, parameter_value in enumerate(parameter_values):
+        metadata = parameter_metadata_dict[index + 1]
+        if metadata.annotations and metadata.annotations.get("ni/type_specialization") == "path":
+            if metadata.repeated:
+                result.append([_path_to_string(value) for value in parameter_value])
+            else:
+                result.append(_path_to_string(parameter_value))
+        else:
+            result.append(parameter_value)
+
+    return result
+
+
+def convert_strings_to_paths(
```

Comment:
This method seems overly complicated.

On initial read, I come away with a few questions:
- `parameter_metadata_dict` is 1-indexed??
- so..., parameter values must be in the order of the id's in the dictionary?, why not have it also be a dict?

Reply:
> parameter_metadata_dict is 1-indexed??

Yes. The keys are protobuf field ids, which are 1-indexed: https://protobuf.dev/programming-guides/proto3/

> so..., parameter values must be in the order of the id's in the dictionary?, why not have it also be a dict?

Good point. That's due to how the underlying serialization code handles parameter values, which is not 100% consistent between configurations (inputs) and outputs.

For configurations (inputs), we do this:
```python
    def _create_measure_request(
        self, parameter_values: List[Any]
    ) -> v2_measurement_service_pb2.MeasureRequest:
        serialized_configuration = any_pb2.Any(
            type_url=${configuration_parameters_type_url | repr},
            value=serialize_parameters(
                parameter_metadata_dict=self._configuration_metadata,
                parameter_values=parameter_values,
                service_name=f"{self._service_class}.Configurations",
            )
        )
        return v2_measurement_service_pb2.MeasureRequest(
            configuration_parameters=serialized_configuration,
            pin_map_context=self._pin_map_context._to_grpc(),
        )
```

This passes the parameters to https://github.com/ni/measurement-plugin-python/blob/main/packages/service/ni_measurement_plugin_sdk_service/_internal/parameter/encoder.py#L17 which expects a `Sequence[Any]`. Originally, this was used on the server to take a `list` or `tuple` returned from the service's `measure()` function and convert it to a gRPC message.

For outputs, we do this (prior to this PR):
```python
    def _deserialize_response(
        self, response: v2_measurement_service_pb2.MeasureResponse
    ) -> Outputs:
        if self._output_metadata:
            result = [None] * max(self._output_metadata.keys())
        else:
            result = []
        output_values = deserialize_parameters(
            self._output_metadata, response.outputs.value, f"{self._service_class}.Outputs"
        )

        for k, v in output_values.items():
            result[k - 1] = v
        return Outputs._make(result)
```

`output_values` is a `dict` indexed by protobuf field id, but we flatten it into a `list` and use the named tuple's `_make` function (which takes a sequence or iterable of named tuple fields).

I could change `convert_strings_to_paths` to operate on a `dict[int, Any]`, but then it would be inconsistent with `convert_paths_to_strings`.

https://github.com/ni/measurement-plugin-python/blob/main/packages/service/ni_measurement_plugin_sdk_service/_internal/parameter/encoder.py#L38 highlights a way to slightly simplify these functions: use `enumerate(parameter_values, start=1)` to let `enumerate` handle the 1-based indexing.

Reply:
I changed the 1-based indexing to use `enumerate(..., start=1)` and renamed the `index` variable to `id`.

I filed https://github.com/ni/measurement-plugin-python/issues/939 about the inconsistency between `Sequence[Any]` and `Dict[int, Any]`.

Reply:
On second thought, the codegen template shouldn't be aware of this conversion. The client_support functions should handle it automatically. I'll change client_support to export functions that do what we need, instead of exporting internal functions that do 95% of what we need.

Reply:
`client_support` now defines its own versions of `deserialize_parameters` and `serialize_parameters` that do path/str conversions and `dict[int, Any]`/`Sequence[Any]` conversions. This hides as much of the complexity as possible from the codegen template.